sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -1517,6 +2231,99 @@ submitted in its entirety for execution via
   ...
 ----
 
+=== Dynamic Parameter Update
+
+Example showing a graph with two kernel nodes, the first created using a kernel
+bundle with `handler::set_args()`, having its node arguments updated.
+
+[source,c++]
+----
+...
+
+using namespace sycl;
+namespace sycl_ext = sycl::ext::oneapi::experimental;
+
+queue myQueue;
+auto myContext = myQueue.get_context();
+auto myDevice = myQueue.get_device();
+
+// USM allocations for kernel input/output
+const size_t n = 1024;
+int *ptrX = malloc_shared<int>(n, myQueue);
+int *ptrY = malloc_device<int>(n, myQueue);
+
+int *ptrZ = malloc_shared<int>(n, myQueue);
+int *ptrQ = malloc_device<int>(n, myQueue);
+
+// Initialize USM allocations
+
+// Kernel loaded from kernel bundle
+const std::vector<kernel_id> builtinKernelIds =
+      myDevice.get_info<info::device::built_in_kernel_ids>();
+kernel_bundle<bundle_state::executable> myBundle =
+      get_kernel_bundle(myContext, { myDevice }, builtinKernelIds);
+kernel builtinKernel = myBundle.get_kernel(builtinKernelIds[0]);
+
+// Graph containing a two kernels node
+sycl_ext::command_graph myGraph(myContext, myDevice);
+
+// Create graph dynamic parameters
+dynamic_parameter<sycl_ext::dynamic_parameter_type::usm> dynParamInput(myGraph);
+dynamic_parameter<sycl_ext::dynamic_parameter_type::usm> dynParamOutput(myGraph);
+dynamic_parameter<sycl_ext::dynamic_parameter_type::scalar> dynParamScalar(myGraph);
+
+// First node uses ptrX as an input & output parameter, with operand
+// mySclar as another argument.
+int myScalar = 42;
+node nodeA = myGraph.add[&](handler& cgh) {
+    cgh.set_args(ptrX, myScalar);
+    cgh.parallel_for(range {n}, builtinKernel);
+});
+
+// Register nodeA dynamic parameters
+dynParamInput.register(0, nodeA);  // Argument index 0 is ptrX
+dynParamInput.register(1, nodeA);  // Argument index 1 is myScalar
+
+// Second node uses ptrX as an inputs, and ptrY as input/output/
+node nodeB = myGraph.add[&](handler& cgh) {
+    cgh.parallel_for(sycl::range<1>{n}, [=](sycl::id<1> it) {
+      const size_t i = it[0];
+      ptrY[i] += ptrX[i];
+    });
+   },
+   {sycl_ext::property::node::depends_on(nodeA)});
+
+// Register nodeB dynamic parameters
+dynParamInput.register(ptrX, nodeB);
+dynParamInput.register(ptrY, nodeB);
```

Comment:
I like the concept of the `dynamic_parameter` type, which acts like a handle to the parameter you want to update.  However, I hate the way the registration works -- matching by value seems error prone.  Could you structure `dynamic_parameter` like this instead?

```
template<typename T>
class dynamic_parameter {
 public:
  dynamic_parameter(const T& param, command_graph<graph_state::modifiable> graph, std::string label = "");
  register(const handler &cgh);
  update(const T& newval, command_graph<graph_state::executable> execGraph);
};
```

Usage would be like this:

```
size_t len = /* ... */;
int *ptrx = malloc_shared<int>(len, myQueue);
int *ptry = malloc_device<int>(len, myQueue);

// Note that we use CTAD here to deduce the template parameter of "dyanmic_parameter".
// Also note that the dynamic parameter is associated with the captured kernel argument in the
// constructor, which seems nice.
dynamic_parameter dparam_len{len};
dynamic_parameter dparam_ptrx{ptrX};
dynamic_parameter_dparam_ptry(ptry};

node nodeB = myGraph.add[&](handler& cgh) {
    // Register each dynamic parameter with the "handler".  The implementation use this to figure out the
    // node.  By the time "myGraph.add" returns, each dynamic parameter becomes associated with "nodeB".
    dparam_len.register(cgh);
    dparam_ptrx.register(cgh);
    dparam_ptry.register(cgh);

    cgh.parallel_for(sycl::range<1>{n}, [=](sycl::id<1> it) {
      const size_t i = it[0];
      ptrY[i] += ptrX[i];
    });
   });

auto execGraph = myGraph.finalize({sycl_ext::property::graph::updatable});
myQueue.ext_oneapi_graph(execGraph);

len = /* some new length */;
int *ptrz = malloc_shared<int>(len, myQueue);
int *ptrq = malloc_device<int>(len, myQueue);

// Update parameter to new values.
dparam_len.update(len, execGraph);
dparam_ptrx.update(ptrz, execGraph);
dparam_ptry.update(ptrq, execGraph);
```

In this proposal, registration is somewhat different.  The `dynamic_parameter` object is constructed with the *address* of the parameter.  The variable passed to the `dynamic_parameter` constructor must be the same variable that is captured by the kernel's lambda.  The call to `register` then associates the dynamic parameter object with the node and the kernel that the node contains.

I'm assuming here that your graph compiler knows the variables that are captured by the kernel.  The graph compiler can search this list of captured variables and compare their *address* with the address that was passed to the `dynamic_parameter` constructor.  This address is unique, so there's no ambiguity with the registration.  You can even create a `dynamic_parameter` for a scalar as my example shows.

Reply:
Thanks for the suggestions. In principal I think having the `dynamic_parameter` be templated on the arg type is fine, though we had initially avoided this to allow generic storage of `dynamic_parameter`s of a single type. This might present a problem for usability but that's my only concern really.

Registration with the handler would present significant problems with record and replay functionality I believe though, since the code may either be not accessible to make these additions (library case) or not desirable for the user to change (general record and replay case). Record and replay is being used in the primary use-case we have for developing this feature against, so compatibility in this area is very important.

I also believe that scalars still present a problem even with this suggestion. It is my understanding that since these parameters are captured by value inside the lambda, when inspected they will have a different address than the original scalar value. Also in cases where `set_arg` is used for scalar values then as far as I can see the `handler` will create it's own copy of the data internally which would have the same issue. Am I misunderstanding something here?

Reply:
> I also believe that scalars still present a problem even with this suggestion. It is my understanding that since these parameters are captured by value inside the lambda, when inspected they will have a different address than the original scalar value. Also in cases where `set_arg` is used for scalar values then as far as I can see the `handler` will create it's own copy of the data internally which would have the same issue. Am I misunderstanding something here?

My proposal is assuming that you have have access to the address of the original variables that are captured by the kernel lambda.  I'm not sure if it's easy for you to get these addresses, but I think it would be worth some extra effort.  The main weakness I see with the original proposal is that it identifies kernel arguments by value, but this is error prone because two arguments might have the same value.  If you could somehow get the address of the captured variables, you can use this address to uniquely associate each `dynamic_parameter` with its kernel argument.

I'm not concerned about the `set_arg` case.  In this case, you can uniquely associate each `dynamic_parameter` with its argument by using the argument index.  The main problem I'm trying to solve in this comment is the case when kernel arguments are captured by the lambda expression.

Do you think you can get access to these addresses somehow?

Reply:
for the case if we could not get node from recorded graph, we can use `void register(void * param);` for USM.

for the case if we can get node from recorded graph, we can use `void register(T param, Node node);` with limitations.

for the case with explicit building API, can we change a bit to `void register(T param, const handler &cgh);`  as below, and we even don't need the address.
```
dynamic_parameter dp(myGraph);
...
node nodeB = myGraph.add[&](handler& cgh) {
    dparam_len.register(len, cgh);
...
```

Reply:
> for the case with explicit building API, can we change a bit to `void register(T param, const handler &cgh);` as below, and we even don't need to care the address.

This doesn't address my concern.  What if the kernel has two parameters that have the same value as `len`?  How will you decide which of those parameters is associated with `dp_len`?

Reply:
how about the below example,

```
int val1 = 11;
int val2 = 22;
dynamic_parameter dparam_val1{sizeof(val1), graph};
dynamic_parameter dparam_val2{sizeof(val2), graph};
dynamic_parameter dparam_ptrx{graph};
dynamic_parameter_dparam_ptry(graph};

node nodeB = myGraph.add[&](handler& cgh) {
    // Register each dynamic parameter with the "handler".  The implementation use this to figure out the
    // node.  By the time "myGraph.add" returns, each dynamic parameter becomes associated with "nodeB".
    dparam_val1.register(val1, cgh);
    dparam_val2.register(val2, cgh);
    dparam_ptrx.register(ptrx, cgh);
    dparam_ptry.register(ptry, cgh);

    cgh.parallel_for(sycl::range<1>{n}, [=](sycl::id<1> it) {
      const size_t i = it[0];
      ptrY[i] += ptrX[i] + val1 + val2;
    });
   });

```

Reply:
In that example, how do these lines register the variable:

```
    dparam_val1.register(val1, cgh);
    dparam_val2.register(val2, cgh);
```

Are they registering the variable by its address (e.g. by taking a reference to `val1`) or are they registering the variable by its value?

Reply:
first of all, I don't know the implementation, just thinking about the possible way.

imo, `dparam_val1.register(val1, cgh); ` maps `dparam_val1` to the kernel(cgh) argument, and the new value will be updated to that argument during `dparam_val1.update(newVal, execGraph);`, and so no address/reference to val1 is required.

Reply:
> Do you think you can get access to these addresses somehow?

I'm not sure how it will be possible to become aware of the addresses of arguments captured in a lambda.

If I understand right, in this example we'll want to connect the `_data` kernel parameter (which we can currently retrieve by runtime queries), to dynamic_parameter `d` which has the address of `data`.

The kernel lambda object is created by the compiler before being passed to any runtime call, and I'm not aware of anything we do at runtime could make the connection to how the compiler transforms the lambda captures (without major compiler changes I couldn't estimate the timeline to deliver).

```cpp
int data = ...;
dynamic_parameter d(data);


auto node = myGraph.add[&](handler& cgh) {
    d.register(cgh);
	
    cgh.parallel_for(sycl::range<1>{n}, [=, _data=data](sycl::id<1> it) {
      const size_t i = it[0];
      ptr[i] = _data;
    });
});
```
Will ask a DPC++ maintainer for an opinion on whether this approach is feasible.

Reply:
Here's another idea.  Does it help if we require the application to capture the `dynamic_parameter` object as the kernel argument?  For example:

```
// Note that we use CTAD here to deduce the template parameter of "dyanmic_parameter".
// Also note that the dynamic parameter is associated with the captured kernel argument in the
// constructor, which seems nice.
dynamic_parameter dparam_len{len};
dynamic_parameter dparam_ptrx{ptrX};
dynamic_parameter_dparam_ptry(ptry};

node nodeB = myGraph.add[&](handler& cgh) {
    // Register each dynamic parameter with the "handler".  The implementation use this to figure out the
    // node.  By the time "myGraph.add" returns, each dynamic parameter becomes associated with "nodeB".
    dparam_len.register(cgh);
    dparam_ptrx.register(cgh);
    dparam_ptry.register(cgh);

    cgh.parallel_for(sycl::range<1>{n}, [=](sycl::id<1> it) {
      const size_t i = it[0];
      dparam_ptry.get()[i] += dparam_ptrx.get()[i];  // Note that "ptrY" and "ptrX" are no longer directly referenced inside the kernel
    });
   });
```

Does this let you capture enough information about the parameters, so that you can uniquely identify a parameter without assuming each parameter has a unique value?

Reply:
I don't see how it's different from a regular buffer/accessor, other than being able to create this new entity from a USM pointer.

Reply:
Sorry, I don't understand your comment.  Did you mean to respond to a different thread in this PR?

Reply:
No to this. I don't see how 

> Does it help if we require the application to capture the dynamic_parameter object as the kernel argument?

would make `dynamic_parameter` substantially different from a buffer/accessor pair. If we have to modify sources in this way, then using proper buffer/accessor instead would solve the issue, wouldn't it?

Reply:
I think `dynamic_parameter` is orthogonal to buffers and accessors.  For example, I think you would still need to use `dynamic_parameter` if you wanted to change the value of an accessor parameter in a graph.  I guess the syntax would look something like:

```
buffer<int> buf{{N}};
accessor acc{buf};

dynamic_parameter dparam_acc{acc};

node nodeB = myGraph.add[&](handler& cgh) {
    cgh.require(acc);
    dparam_acc.register(cgh);

    cgh.parallel_for(sycl::range<1>{n}, [=](sycl::id<1> it) {
      const size_t i = it[0];
      dparam_acc.get()[i] += i;
    });
   });
```

Reply:
I was confused by the first example that only used scalar `len` variable and USM pointers. Do we expect that "updated" accessor value would have all the same compile-time properties between old/new values (e.g. number of dimensions)?

Reply:
> Do we expect that "updated" accessor value would have all the same compile-time properties between old/new values (e.g. number of dimensions)?

I assume so, but this is a question to the PR authors.

Assuming the updated accessor *does* need to have the same compile-time properties, I think the `dynamic_parameter` would be able to enforce this because those compile-time properties would become part of the type of the `dynamic_parameter` object.  The `dynamic_parameter` class is templated on the type of the argument, so we could enforce (at compile time) that the updated argument value had the same type (which would include accessor properties).

Reply:
> Does this let you capture enough information about the parameters, so that you can uniquely identify a parameter without assuming each parameter has a unique value?

Your proposal of a `dynamic_parameter` instance crossing the host/device code boundary I think should let us capture enough information to uniquely identify each parameter, as the `dynamic_parameter` instance itself becomes the parameter. 

My main concern with this design is that it requires the user to modify the kernel code, which won't work for the record & replay usecase where existing code is wrapped. If I understand @guoyejun's description of how pytorch works, this will be their usage before updating a graph instance. What we could do is aim to provide **both** registration mechanisms, matching-by-value and the CGH registration, then try to communicate to the user in the spec the drawbacks of the match-by-value approach.

I'm also not sure how long it will take to implement dynamic_parameters as a cross host/device construct, there's
a lot more uncertainty in terms of implementation. E.g. Ensure that when a user has multiple instances of the same graph,
and updates only one of them, then dynamic_parameter.get() has been implemented correctly so that only kernels in that exec graph instance are updated. So having both mechanisms means we could at least prototype match-by-value for evaluation on an estimatable timescale and get feedback on how often match-by-value collisions happen in real workloads.

Reply:
@EwanC yes, that's right.

to be accurate, PyTorch is not currently using the graph update, but the future refinement in our discussion with pytorch people.

Currently in PyTorch, there are two cases if the graph input USM is changed, 
1)  add an extra copy (in pytorch eager mode)
2)  re-record the graph (in torch.compile mode)